### ISSUE TITLE
View constraints which led a state to be unsat

### ISSUE THREAD

**Issue by jaiverma** [angr/angr#2321]
I was trying to find opaque predicates. This is a basic-block which is an opaque predicate:

```
>>> block.pp()
0x491bc9:       mov     dword ptr [esi + 0x3c], ebx
0x491bcc:       mov     edi, dword ptr [0x590d7c]
0x491bd2:       mov     ebx, dword ptr [0x590d78]
0x491bd8:       mov     eax, edi
0x491bda:       imul    eax, eax
0x491bdd:       imul    eax, eax, 7
0x491be0:       dec     eax
0x491be1:       imul    ebx, ebx
0x491be4:       cmp     eax, ebx
0x491be6:       je      0x491bf5
```

The `cmp eax, ebx` instruction just before the conditional jump will always be false since the condition is `unsat`.

The constraints are:
`eax * eax * 7 - 1 == ebx * ebx`

I am trying to get the constraints using the following snippet:

```py
import angr
import claripy

proj = angr.Project('samples/ac3e087e43be67bdc674747c665b46c2')
block = proj.factory.block(0x00491bc9)

state = proj.factory.blank_state(addr=0x00491bc9)
# state.regs.eax = claripy.BVS('eax', 32)
# state.regs.ebx = claripy.BVS('ebx', 32)

# using save_unsat to store unsat states in unsat stash
# instead of discarding them
simgr = proj.factory.simulation_manager(state, save_unsat=True)

simgr.step()

print(simgr.stashes)

x = simgr.stashes['active'][0]
y = simgr.stashes['unsat'][0]

print(x.solver.constraints)
print(y.solver.constraints)
```

The problem is, I’m getting the same constraints for both the state in the `active` stash as well as the one in the `unsat` stash…

```py
defaultdict(<class 'list'>, {'active': [<SimState @ 0x491be8>], 'stashed': [], 'pruned': [], 'unsat': [<SimState @ 0x491bf5>], 'errored': [], 'deadended': [], 'unconstrained': []})
[<Bool reg_esi_0_32{UNINITIALIZED} + 0x3c <= 0xffffffff>, <Bool reg_esi_0_32{UNINITIALIZED} + 0x3c == 0xffffffff>]
[<Bool reg_esi_0_32{UNINITIALIZED} + 0x3c <= 0xffffffff>, <Bool reg_esi_0_32{UNINITIALIZED} + 0x3c == 0xffffffff>]
```

My question is, how can I get the constraints which caused the `y` state to go into the `unsat` stash?

**Comment by ltfish** [comment]
Enable `unsat_core` in claripy and see what the unsat core is for the unsat state solver.

**Comment by jaiverma** [comment]
I enabled `CONSTRAINT_TRACKING_IN_SOLVER` with:
```py
state.options |= {angr.sim_options.CONSTRAINT_TRACKING_IN_SOLVER}
```

And then used the following to print the unsat core:
```py
print(y.solver.unsat_core())
```

But the output is an empty list.
```py
defaultdict(<class 'list'>, {'active': [<SimState @ 0x491be8>], 'stashed': [], 'pruned': [], 'unsat': [<SimState @ 0x491bf5>], 'errored': [], 'deadended': [], 'unconstrained': []})
[<Bool reg_esi_0_32{UNINITIALIZED} + 0x3c <= 0xffffffff>, <Bool reg_esi_0_32{UNINITIALIZED} + 0x3c == 0xffffffff>]
[<Bool reg_esi_0_32{UNINITIALIZED} + 0x3c <= 0xffffffff>, <Bool reg_esi_0_32{UNINITIALIZED} + 0x3c == 0xffffffff>]
[]
```

I also tried single stepping through the basic block and checking the `unsat` stash after every step, but same result. I guess the state goes into the `unsat` stash at the conditional jump anyway because that's when it needs to evaluate the constraints.
```py
for i in range(block.instructions):
    simgr.step(num_inst=1)
    if len(simgr.stashes['unsat']) > 0:
        x = simgr.stashes['unsat'][0]
        print(x.solver.constraints)
        print(x.solver.unsat_core())
```

This also gives an empty list as output.

**Comment by ltfish** [comment]
This is interesting since we are not doing anything fancy with the unsat_core: We directly call into Z3 and returns what Z3 believes is the unsat core. Can you disable composite solver (remove the `COMPOSITE_SOLVER` state option) and try again?

**Comment by jaiverma** [comment]
I disabled `COMPOSITE_SOLVER` with:
```py
state.options -= {angr.sim_options.COMPOSITE_SOLVER}
```

This gives an additional state in the constraints (`<Bool False>`), but `unsat_core` is still empty.
```py
[<Bool reg_esi_0_32{UNINITIALIZED} + 0x3c <= 0xffffffff>, <Bool reg_esi_0_32{UNINITIALIZED} + 0x3c == 0xffffffff>, <Bool False>]
()
```

**Comment by jaiverma** [comment]
I was able to make some progress with this.
Looking at the second and third instructions in this basic block:
```
0x491bcc:       mov     edi, dword ptr [0x590d7c]
0x491bd2:       mov     ebx, dword ptr [0x590d78]
```
I explicitly set these two addresses as:
```py
state.memory.store(0x00590d7c, claripy.BVS('a', 32))
state.memory.store(0x00590d78, claripy.BVS('b', 32))
```
After this, I am able to view the constraints for both the `sat` and `unsat` states at the conditional jump with
```py
print(x.solver.constraints)
print(y.solver.constraints)
```
```py
[<Bool Reverse(a_0_32) * Reverse(a_0_32) * 0x7 - 0x1 != Reverse(b_1_32) * Reverse(b_1_32)>]
[<Bool Reverse(a_0_32) * Reverse(a_0_32) * 0x7 - 0x1 == Reverse(b_1_32) * Reverse(b_1_32)>]
```

**Comment by jaiverma** [comment]
There is an exception raised though when i run `unsat_core()` for state `y`.
```py
>>> y.solver.satisfiable()                                                                                                                                    
False                                                                                                                                                         
>>>                                                                                                                                                           
>>> y.solver.unsat_core()                                                                                                                                     
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jai/.venvs/angr/lib/python3.8/site-packages/angr/state_plugins/sim_action_object.py", line 57, in ast_stripper
    return f(*new_args, **new_kwargs)
  File "/Users/jai/.venvs/angr/lib/python3.8/site-packages/angr/state_plugins/solver.py", line 89, in wrapped_f
    return f(*args, **kwargs)
  File "/Users/jai/.venvs/angr/lib/python3.8/site-packages/angr/state_plugins/solver.py", line 622, in unsat_core
    return self._solver.unsat_core(extra_constraints=extra_constraints)
  File "/Users/jai/.venvs/angr/lib/python3.8/site-packages/claripy/frontends/composite_frontend.py", line 368, in unsat_core
    cores.extend(list(solver.unsat_core(extra_constraints=extra_constraints)))
  File "/Users/jai/.venvs/angr/lib/python3.8/site-packages/claripy/frontends/full_frontend.py", line 210, in unsat_core
    unsat_core = self._solver_backend.unsat_core(self._get_solver())
  File "/Users/jai/.venvs/angr/lib/python3.8/site-packages/claripy/backends/__init__.py", line 466, in unsat_core
    return [ self._abstract(core) for core in self._unsat_core(s) ]
  File "/Users/jai/.venvs/angr/lib/python3.8/site-packages/claripy/backends/__init__.py", line 466, in <listcomp>
    return [ self._abstract(core) for core in self._unsat_core(s) ]
  File "/Users/jai/.venvs/angr/lib/python3.8/site-packages/claripy/backends/backend_z3.py", line 70, in z3_condom
    return f(*args, **kwargs)
  File "/Users/jai/.venvs/angr/lib/python3.8/site-packages/claripy/backends/backend_z3.py", line 393, in _abstract
    return self._abstract_internal(e.ctx.ctx, e.ast)
AttributeError: 'NoneType' object has no attribute 'ctx'
```
This happens regardless of whether I disable `COMPOSITE_SOLVER`.

**Comment by rhelmot** [comment]
that's an error that I've seen show up sporadically in our CI logs, I think it's nondeterministic and has something to do with garbage collection patterns. I'm not super sure what to do about it...

**Comment by jaiverma** [comment]
`self._hash_to_constraint` is a dictionary which stores `name` and `constraint`. This is in file `claripy/backends/backend_z3.py`.
```py
 720     def _add(self, s, c, track=False):                                          
 721         if track:                                                               
 722             for constraint in c:                                                
 723                 name = str(hash(constraint))                                    
 724                 if name not in self._hash_to_constraint:                        
 725                     self._hash_to_constraint[name] = constraint                 
 728                     s.assert_and_track(constraint, name) 
```
This is what happens when we call `simgr.step()`. The `_add` function is called twice.
```py
ipdb> next
> /Users/jai/.venvs/angr/lib/python3.8/site-packages/claripy/backends/backend_z3.py(724)_add()
    723                     self._hash_to_constraint[name] = constraint
--> 724                     s.assert_and_track(constraint, name)
    725         else:

ipdb> print(name)
361508893
ipdb> print(constraint)
Concat(Concat(Concat(Extract(7, 0, a_0_32),
                     Extract(15, 8, a_0_32)),
              Extract(23, 16, a_0_32)),
       Extract(31, 24, a_0_32))*
Concat(Concat(Concat(Extract(7, 0, a_0_32),
                     Extract(15, 8, a_0_32)),
              Extract(23, 16, a_0_32)),
       Extract(31, 24, a_0_32))*
7 -
1 ==
Concat(Concat(Concat(Extract(7, 0, b_1_32),
                     Extract(15, 8, b_1_32)),
              Extract(23, 16, b_1_32)),
       Extract(31, 24, b_1_32))*
Concat(Concat(Concat(Extract(7, 0, b_1_32),
                     Extract(15, 8, b_1_32)),
              Extract(23, 16, b_1_32)),
       Extract(31, 24, b_1_32))
```
The second time, it is called with:
```py
ipdb> c
> /Users/jai/.venvs/angr/lib/python3.8/site-packages/claripy/backends/backend_z3.py(728)_add()
    727                     pdb.set_trace()
--> 728                     s.assert_and_track(constraint, name)
    729         else:

ipdb> print(name)
1513186725
ipdb> print(constraint)
Concat(Concat(Concat(Extract(7, 0, a_0_32),
                     Extract(15, 8, a_0_32)),
              Extract(23, 16, a_0_32)),
       Extract(31, 24, a_0_32))*
Concat(Concat(Concat(Extract(7, 0, a_0_32),
                     Extract(15, 8, a_0_32)),
              Extract(23, 16, a_0_32)),
       Extract(31, 24, a_0_32))*
7 -
1 !=
Concat(Concat(Concat(Extract(7, 0, b_1_32),
                     Extract(15, 8, b_1_32)),
              Extract(23, 16, b_1_32)),
       Extract(31, 24, b_1_32))*
Concat(Concat(Concat(Extract(7, 0, b_1_32),
                     Extract(15, 8, b_1_32)),
              Extract(23, 16, b_1_32)),
       Extract(31, 24, b_1_32))
```
The problem though is that, `self._hash_to_constraint_clear()` is called before each `_add`, which causes `self._hash_to_constraint` to only contain an entry for `1513186725`.
```py
ipdb> c
> /Users/jai/.venvs/angr/lib/python3.8/site-packages/claripy/backends/backend_z3.py(709)solver()
    708         pdb.set_trace()
--> 709         self._hash_to_constraint.clear()
    710 
```
So, eventually, the error occurs when `_unsat_core` in `backend_z3.py` is called:
```py
741     def _unsat_core(self, s):                                                   
 742         cores = s.unsat_core()                                                  
 743         constraints = [ ]                                                       
 744         for core in cores:                                                      
 745             constraints.append(self._hash_to_constraint.get(str(core)))         
 746         return constraints 
```
Here `s.unsat_core()` returns:
```py
ipdb> up
> /Users/jai/.venvs/angr/lib/python3.8/site-packages/claripy/backends/__init__.py(466)unsat_core()
    464         """
    465 
--> 466         return [ self._abstract(core) for core in self._unsat_core(s) ]
    467 
    468     def _unsat_core(self, s):  #pylint:disable=no-self-use,unused-argument

ipdb> print(s.unsat_core())
[361508893]
```
Since `_hash_to_constraint` was cleared and now only contains `1513186725`, `self._abstract` is called with `None` which causes the exception.

**Comment by jaiverma** [comment]
Interestingly, everything works if I single step each instruction in the basic-block instead of doing `simgr.step()` to execute the whole block.
```py
for i in range(block.instructions):
    simgr.step(num_inst=1)
    if len(simgr.stashes['unsat']) > 0:
        x = simgr.stashes['unsat'][0]
        print(x.solver.unsat_core())
```
The `unsat_core` is:
```py
(<Bool 0xffffffff + 0x7 * ((a_0_32[7:0] .. a_0_32[15:8] .. a_0_32[23:16] .. a_0_32[31:24]) * (a_0_32[7:0] .. a_0_32[15:8] .. a_0_32[23:16] .. a_0_32[31:24])) - (b_1_32[7:0] .. b_1_32[15:8] .. b_1_32[23:16] .. b_1_32[31:24]) * (b_1_32[7:0] .. b_1_32[15:8] .. b_1_32[23:16] .. b_1_32[31:24]) == 0x0>,)
```
Any tips on how I can go about fixing/debugging this further?

**Comment by ltfish** [comment]
I think the is the exact issue that I debugged in the past and was convinced that Z3's Python binding was the culprit. I could not trigger it reliably though. Does your script trigger it reliably?

**Comment by jaiverma** [comment]
Yes, exception occurs every single time.

Here is the complete script by the way: [solve.py](https://gist.github.com/jaiverma/0817b7f4a8a42ca7294e6fb853fec5b6)
This is the binary which I'm using: [ac3e087e43be67bdc674747c665b46c2](https://github.com/mrphrazer/r2con2020_deobfuscation/blob/master/samples/ac3e087e43be67bdc674747c665b46c2)

**Comment by r4dr3fr4d** [comment]
Thought I'd check in on this.   Every now and then I have a want to see what's specifically unsatisfiable and every time unsat_core() returns an empty list for me as well.

**Comment by rhelmot** [comment]
Should be fixed by angr/claripy#220 :D
